charm_temp_errors.md

```diff
@@ -0,0 +1,155 @@
+
+> :warning: **Don't use this code** - there's a better alternative below!
+
+```python
+if not container.can_connect():
+    event.defer()
+    return
+try:
+    container.replan()
+except ops.pebble.ConnectionError:
+    event.defer()
+    return
+```
+
+However, it's clear now that the `can_connect()` call isn't doing anything useful - it's pinging Pebble but most of the time we expect Pebble will be available (so it's not required) and when it's not we already have the `try`/`except` to handle that. Let's make an attempt to simplify the code:
+
+```python
+try:
+    container.replan()
+except ops.pebble.ConnectionError:
+    event.defer()
+    return
+```
+
+We're pretty sure that when we do get this failure, it's a temporary problem - perhaps there's high load, or some temporary networking issues. One issue with using `defer()` is that we don't know, or have any control over, when the event handler will be retried. `defer()` also restarts the entire handler, so any work we've already done will get repeated, which is wasteful. Let's adjust the code so that we give Pebble a few chances:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+        break
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY * attempt)
+else:
+    event.defer()
+    return
+```
+
+[tenacity](https://tenacity.readthedocs.io/en/latest/) is a really nice library providing decorators and context managers that handle different backoff mechanisms, attempt limits, and so forth - but if you want to keep your dependencies to a minimum, it's simple enough to have something like the above as reusable code within your charm.
+
+At this point, if Pebble is responding, then we'll get our `replan()` executed in essentially the same time as if we just called it (certainly faster than if we guarded with `can_connect()`), and if there's an issue with Pebble, we'll retry until we're confident that the problem isn't going to resolve itself within a timeframe we're comfortable with (perhaps a couple of minutes - remembering that the Charm won't handle other events while it's sleeping between retries). If Pebble is failing longer than that, do we really want to defer? That would be appropriate if we believe that the issue will resolve itself, but if Pebble has been unreachable this long, it's likely that someone needs to investigate what's happening - ie. manual intervention is required, and a Juju error state is appropriate in that case. Let's adjust the code again:
+
+```python
+for attempt in range(MAX_PEBBLE_RETRIES):
+    try:
+        container.replan()
+        break
+    except ops.pebble.ConnectionError:
+        time.sleep(PEBBLE_RETRY_DELAY * attempt)
+else:
+    raise RuntimeError("Unable to reach Pebble")
```

Comment:
With this approach we lose the original exception. Could we instead re-raise the exception, or at least include the original exception, in the final one?

```python
for attempt in range(MAX_PEBBLE_RETRIES):
    try:
        container.replan()
        break
    except ops.pebble.ConnectionError:
        if attempt == MAX_PEBBLE_RETRIES - 1:
            raise
        time.sleep(PEBBLE_RETRY_DELAY * attempt)
```